### PR TITLE
refactor(*): Improve scene objects cohesion

### PR DESCRIPTION
**Pull request by grafakus**
### ✨ Description

**Related issue(s):** is blocked by https://github.com/grafana/explore-profiles/pull/80 (should be merged after)

This PR pulls out specific Scene objects responsibilities out of the root `SceneProfilesExplorer` class. We do this in order to improve the cohesion of each Scene object responsible for a type of exploration. Eventually we hope the code will be easier to maintain and to reason about.

### 📖 Summary of the changes

There are 2 main changes:
1. Each Scene object responsible for an exploration type tells `SceneProfilesExplorer` which variables and grid controls to render
2. For events emitted as a result of a user click on a panel action, `SceneProfilesExplorer` now passes them to the Scene object constructor, which responsibility is to initialize properly the variables it uses

See diff tab for specific comments.

### 🧪 How to test?

- The build should pass

### CODE REVIEW

File: src/pages/ProfilesExplorerView/components/SceneExploreAllServices/SceneExploreAllServices.tsx [hunk 1]
```diff
@@ -35,6 +39,24 @@ export class SceneExploreAllServices extends SceneObjectBase<SceneExploreAllServ
...
+    this.addActivationHandler(this.onActivate.bind(this));
+  }
+
+  onActivate() {
+    const quickFilter = findSceneObjectByClass(this, SceneQuickFilter) as SceneQuickFilter;
+    quickFilter.setPlaceholder('Search services (comma-separated regexes are supported)');
+  }
+
+  // see SceneProfilesExplorer
+  getVariablesAndGridControls() {
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreAllServices/SceneExploreAllServices.tsx:52]: The Scene object now tells `SceneProfilesExplorer` which variables to render.

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneExploreServiceFlameGraph.tsx [hunk 2]
```diff
@@ -29,6 +34,56 @@ export class SceneExploreServiceFlameGraph extends SceneObjectBase<SceneExploreS
       }),
       body: new SceneFlameGraph(),
     });
+
+    this.addActivationHandler(this.onActivate.bind(this, item));
+  }
+
+  onActivate(item?: GridItemData) {
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneExploreServiceFlameGraph.tsx:41]: When it receives an item (which happens in runtime when a user clicks on a panel action), the Scene object is now responsible for initializing properly the variables it uses.

File: src/pages/ProfilesExplorerView/components/SceneExploreFavorites/SceneExploreFavorites.tsx [hunk 3]
```diff
@@ -56,6 +60,9 @@ export class SceneExploreFavorites extends SceneObjectBase<SceneExploreFavorites
   }
 
   onActivate() {
+    const quickFilter = findSceneObjectByClass(this, SceneQuickFilter) as SceneQuickFilter;
+    quickFilter.setPlaceholder('Search favorites (comma-separated regexes are supported)');
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreFavorites/SceneExploreFavorites.tsx:64]: If it uses the quick filter, the Scene object also sets the input placeholder

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneExploreServiceFlameGraph.tsx [hunk 4]
```diff
@@ -29,6 +34,56 @@ export class SceneExploreServiceFlameGraph extends SceneObjectBase<SceneExploreS
...
+    if (item) {
+      this.initVariables(item);
+    }
+
+    const profileMetricVariable = findSceneObjectByClass(this, ProfileMetricVariable) as ProfileMetricVariable;
+
+    profileMetricVariable.setState({ query: ProfileMetricVariable.QUERY_SERVICE_NAME_DEPENDENT });
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneExploreServiceFlameGraph.tsx:48]: The Scene object is also responsible for its own specific setup/teardown (like making the profile metric variable aware of the selected service)

File: src/pages/ProfilesExplorerView/components/SceneProfilesExplorer/SceneProfilesExplorer.tsx [hunk 5]
```diff
@@ -1,11 +1,9 @@
 import { css } from '@emotion/css';
 import { dateTimeParse, GrafanaTheme2 } from '@grafana/data';
 import {
-  DataSourceVariable,
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneProfilesExplorer/SceneProfilesExplorer.tsx:4]: The main goal of the PR was to remove as much code as possible from this class.

File: src/pages/ProfilesExplorerView/infrastructure/timeseries/buildTimeSeriesQueryRunner.ts [hunk 6]
```diff
@@ -3,7 +3,6 @@ import { SceneQueryRunner } from '@grafana/scenes';
 import { PYROSCOPE_DATA_SOURCE } from '../pyroscope-data-sources';
 import { TimeSeriesQueryRunnerParams } from './TimeSeriesQueryRunnerParams';
 
-// eslint-disable-next-line sonarjs/cognitive-complexity
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/infrastructure/timeseries/buildTimeSeriesQueryRunner.ts:6]: Not directly related to this PR but removing an `eslint-disable-next-line`. cannot be bad :)

File: src/pages/ProfilesExplorerView/components/SceneProfilesExplorer/SceneProfilesExplorer.tsx [hunk 7]
```diff
@@ -419,9 +327,12 @@ export class SceneProfilesExplorer extends SceneObjectBase<SceneProfilesExplorer
     const { explorationType, controls, body } = this.useState();
 
     const [timePickerControl, refreshPickerControl] = controls as [SceneObject, SceneObject];
+    const dataSourceVariable = this.state.$variables!.state!.variables[0] as ProfilesDataSourceVariable;
 
-    const { variables, gridControls } = this.getVariablesAndGridControls(explorationType as ExplorationType);
-    const [dataSourceVariable, ...sceneVariables] = variables as SceneVariable[];
+    const { variables: sceneVariables, gridControls } = (body?.state.primary as any).getVariablesAndGridControls() as {
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneProfilesExplorer/SceneProfilesExplorer.tsx:332]: TODO in a future PR: enforce `getVariablesAndGridControls` to be a method that exists and returns the adequate object.

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceLabels/SceneExploreServiceLabels.tsx [hunk 8]
```diff
@@ -47,6 +54,78 @@ export class SceneExploreServiceLabels extends SceneObjectBase<SceneExploreServi
...
+      // we have to wait until the new groupBy options have been loaded
+      // if not, its value will default to "all" regardless of our call to "changeValueTo"
+      // this happens, e.g., when landing on Favorites then jumping to "Service label" by clicking on a favorite that contains a "groupBy" label value
+      const groupBySub = groupByVariable.subscribeToState((newState, prevState) => {
+        if (!newState.loading && prevState.loading) {
+          groupByVariable.changeValueTo(groupBy.label);
+          groupBySub.unsubscribe();
+        }
+      });
+    }
```
Comment (bryanhuhta) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceLabels/SceneExploreServiceLabels.tsx:111]: 👍

**Review by bryanhuhta — DISMISSED** [review]

**Review by bryanhuhta — APPROVED** [review]

**Pull request merged**